docs/standard/library-guidelines/dependencies.md

```diff
@@ -0,0 +1,95 @@
+```
+
+The rules that NuGet uses when resolving dependencies are [complex](https://docs.microsoft.com/en-us/nuget/consume-packages/dependency-resolution), but NuGet always attempts to find the lowest applicable version. NuGet prefers the lowest application version over using the highest available because the lowest will have the least compatibility issues.
+
+Because of NuGet's lowest application version rule, it is not necessary to place an upper version or exact range on package references to avoid getting the latest version. NuGet already tries to find the lowest, most compatible version for you.
+
+```xml
+<!-- Accepts 1.0 up to 1.x, but not 2.0 and higher. -->
+<PackageReference Include="ExamplePackage" Version="[1.0,2.0)" />
+
+<!-- Accepts exactly 1.0. -->
+<PackageReference Include="ExamplePackage" Version="[1.0]" />
+```
+
+Upper version limits will cause NuGet to fail if there is a conflict, e.g. one library accepts exactly 1.0 while another library requires 2.0 or above. While breaking changes may have been introduced in version 2.0, a strict or upper limit version dependency guarantees an error.
+
+![Diamond dependency conflict](./media/diamond-dependency-conflict.png "Diamond dependency conflict")
+
+**❌ DO NOT** have NuGet package references with no minimum version.
+
+**❌ AVOID** NuGet package references that demand an exact version.
+
+**❌ AVOID** NuGet package references with a version upper limit.
+
+**More Information**
+
+* [How NuGet resolves package dependencies](https://docs.microsoft.com/en-us/nuget/consume-packages/dependency-resolution)
+
+## NuGet shared source packages
+
+One way to reduce external NuGet package dependencies is to reference share source packages. A shared source package contains source code files that are included in a project when referenced. Because you are just including source code files that are compiled with the rest of your project there is no external dependency and chance of conflict.
+
+Shared source packages are great for including small pieces of functionality, e.g. a shared source package containing helper methods for making HTTP calls.
+
+![Shared source package](./media/shared-source-package.png "Shared source package")
+
+Shared source packages can only be used by `PackageReference`, and should be a private reference to tell NuGet it is only used at development time and should not be exposed to anyone using your package.
+
+```xml
+<PackageReference Include="Microsoft.Extensions.Buffers.Testing.Sources" PrivateAssets="All" Version="1.0" />
+```
+
+![Shared source project](./media/shared-source-project.png "Shared source project")
+
+**✔️ CONSIDER** referencing shared source packages for small, internal pieces of functionality.
+
+**✔️ CONSIDER** making your package a shared source package if it provides small, internal pieces of functionality.
+
+**✔️ DO** reference shared source packages with `PrivateAssets="All"`.
+
+**❌ AVOID** shared-source package types in your public API.
```

Comment:
I'd go further and make this a "DO NOT" since I can't think of a case where it would actually work...

Reply:
Well, it does work for small things like a command line parser but it has severe drawbacks:

1. It's language specific (and thus in practice excludes VB & F#)
2. It doesn't unify (and thus should only be used with `internal` modifier on types

I agree with @anurse that we should make this a **DO NOT**.

Reply:
I guess that's another way to put it:

> **❌ DO NOT** push shared-source packages to nuget.org.
> > Shared-source packages (i.e. packages containing source code rather than a library) do not work well for multiple languages and will cause conflicts if containing types are used in public APIs. Consider these internal infrastructure building blocks, rather than packages that are suitable for publication.

Reply:
That's not 100% the case, there are cases that can work as published shared-source packages (i.e. [LibLog](https://www.nuget.org/packages/LibLog)). I think it's reasonable to tell users that they really better be sure they know what they are doing though, and a `DO NOT` does pretty good there.

Reply:
I think an IMPORTANT section to highlight this limitation of source packages would be good. Developers can then judge for themselves whether distributing source is the right for them.

Reply:
I can't tell whether you agree with my proposed rewording then or disagree.

LibLog is actually a great example of the problem: it only works in C#, VB and F# users are left hanging.

Reply:
> The OSS author can use that info to judge whether distributing source is the right for their project.

Maybe. At the same time, the value of these guidlines is to highlight best practices. And due to the problems I think it's fair to say that publishing source isn't a best practice, which justifies a **DO NOT** guideline. However, these guidelines aren't binding -- they are simply the accumulation of our experience. But that doesn't mean we should take the stance of giving developers a bunch of facts and tell them "decide for yourself".

Here is another way to think about: what guidance would we want to give first party teams, such as Azure SDK. I'd think we'd tell them "don't do source packages".

Reply:
Assuming the goal of this guidance is to describe what *great* .NET libraries for the entire ecosystems look like, I'd say I do feel strongly that we should say **DO NOT** 😄 

> We would nee to reword the paragraphs above the recommendations to talk about this limitation. 

Agreed. Do think this isn't enough?

> **❌ DO NOT** push shared-source packages to nuget.org.
> > Shared-source packages (i.e. packages containing source code rather than libraries) do not work well for multiple languages and will cause conflicts if containing types are used in public APIs. Consider these internal infrastructure building blocks, rather than packages that are suitable for publication.